server/src/main/java/org/elasticsearch/common/util/BigArrays.java

```diff
@@ -356,6 +356,88 @@ public T set(long index, T value) {
 
     }
 
+    abstract public static class BinarySearcher<T>{
+
+        /**
+         * @return a negative integer, zero, or a positive integer if the
+         * array's value at <code>index</code> is less than, equal to, or greater than <code>value</code>
+         */
+        abstract protected int compare(int index, T value);
+
+        /**
+         * @return the magnitude of the distance between <code>value</code> and the element at <code>index</code>
+         * It will usually be <code>Math.abs(array[index] - value)</code>
+         */
+        abstract protected double distance(int index, T value);
+
+        /**
+         * @return the index who's underlying value is closest to <code>value</code>
+         */
+        private int getClosestIndex(int index1, int index2, T value){
+            if(distance(index1, value) < distance(index2, value)){
+                return index1;
+            } else {
+                return index2;
+            }
+        }
+
+        /**
+         * Uses a binary search to determine the index of the closest element to <code>value</code> that lies in
+         * the index range {from, ... , to}
+         *
+         * @return the index of the closest element
+         *
+         * Requires: The underlying array should be sorted
+         **/
+        public int search(int from, int to, T value){
+            while(from < to){
+                int mid = (from + to) / 2;
```

Comment:
I *think* for very large bounds this'll overflow and make garbage. You could test it with something like:
```
new BinarySearcher() {
  @Override
  protected int compare(int index, T value) { return Integer.compareTo(index, value);} 
  @Override
  protected int distance(int index, T value) { return Math.abs(index - value);} 
}.search(0, Integer.MAX_VALUE, Integer.MAX_VALUE -10);
```